### PR TITLE
completed fit and created test script for fit 

### PR DESCRIPTION
**Pull request by saurav193**
Please try the fit with as much edge cases as possible and let me know where it fails

### CODE REVIEW

File: Kmeans_python/fit.py [hunk 1]
```diff
@@ -1,7 +1,83 @@
...
+def compute_distance(samples, centers):
+    
+    """
+    This computes the euclidean distance of each sample from all the cluster centers
+    
+    Parameters
+    ----------
+    
+    samples : numpy.ndarray
+    all the data points in our sample
```
Comment (sreejithmunthikodu) [inline Kmeans_python/fit.py]:
```suggestion
    all the data points in the sample
```

File: Kmeans_python/fit.py [hunk 2]
```diff
@@ -1,7 +1,83 @@
...
+                                            np.random.random(n_features)*(max_features - min_features)/3 +  \
+                                                np.random.random(n_features)*(max_features - min_features)/3
+
+    return centers
+
+def fit(X_train, k, n_init = 10, max_iter = 200):
+    """
+    This function classifies the non-labeled data into a given number of clusters k using simple KMeans algorithm.
+    It returns labels for each data point according to the cluster it belongs and also 
     cluster centers. This is a type of unsupervised learning method to classify data.
```
Comment (sreejithmunthikodu) [inline Kmeans_python/fit.py:81]:
```suggestion
    """
    This function classifies the non-labeled data into a given number of clusters k using a simple KMeans algorithm.
    It returns labels for each data point according to the cluster it belongs to and also cluster centers. This is a type of unsupervised learning method to cluster data.
```

File: Kmeans_python/fit.py [hunk 3]
```diff
@@ -30,5 +106,64 @@ def fit(X_train, k):
     >>> centers, labels = fit(X, 2)
     
     """
+    if not ((isinstance(X_train, pd.DataFrame)) | (isinstance(X_train, np.ndarray))):
+        raise ValueError("Invalid input type for samples."
+                         " X_train must be pandas dataframe or a numpy array.")
+        
+    if n_init <= 0:
+        raise ValueError("Invalid number of initializations."
+                         " n_init=%d must be bigger than zero." % n_init)
```
Comment (sreejithmunthikodu) [inline Kmeans_python/fit.py:115]:
```suggestion
                         " n_init=%d must be larger than zero." % n_init)
```

File: Kmeans_python/fit.py [hunk 4]
```diff
@@ -30,5 +106,64 @@ def fit(X_train, k):
...
+
+    if max_iter <= 0:
+        raise ValueError("Number of interations should be a positive integer."
+                         " max_iter=%d must be bigger than zero." % max_iter)
```
Comment (sreejithmunthikodu) [inline Kmeans_python/fit.py:119]:
```suggestion
                         " max_iter=%d must be larger than zero." % max_iter)
```

File: Kmeans_python/fit.py [hunk 5]
```diff
@@ -30,5 +106,64 @@ def fit(X_train, k):
...
+    
+    if k < 1:
+        raise ValueError("Invalid number of clusters."
+                         " k=%d must be an integer bigger than or equal to 1." % k)
```
Comment (sreejithmunthikodu) [inline Kmeans_python/fit.py:123]:
```suggestion
                         " k=%d must be an integer greater than or equal to 1." % k)
```

File: Kmeans_python/fit.py [hunk 6]
```diff
@@ -30,5 +106,64 @@ def fit(X_train, k):
...
+
+            labels = np.argmin(centroid_distances, axis = 1)
+            
+            # updating the centers with mean
+            for m in range(k):
+                
+                centers[m,:] = np.mean(X[np.where(labels == m)[0], : ], axis = 0)
+        
+        updated_inertia = np.sum(np.square(centroid_distances))
+
```
Comment (sreejithmunthikodu) [inline Kmeans_python/fit.py]: I believe, inertia is not the sum of the squared distance of all samples to all cluster centers. But the sum of the squared distance of all samples to their respective cluster centers (sum of intra-cluster squared distance)

```suggestion
            inertia_array = np.zeros((k, 1))
            for m in range(k):                
                centers[m,:] = np.mean(X[np.where(labels == m)[0], : ], axis = 0)
                distance_m = np.min(centroid_distances[np.where(labels == m)[0], : ], axis = 0)
                inertia_array[m, 0] = np.sum(np.square(distance_m))
        
        # Add all intra-cluster distances to get the inertia
        updated_inertia = np.sum(inertia_array)

```

**Review by sreejithmunthikodu — CHANGES_REQUESTED** [review]
Some minor modifications in the docstrings.
I modified the code slightly where the inertia is calculated. I did not run the code. Please check the changes and confirm if the changes make sense.

File: Kmeans_python/fit.py (hunk 2 above)
Comment (saurav193) [reply Kmeans_python/fit.py:81]: Is the change only in the <return> character in the second sentence? If yes, I don't think it needs to be changed

File: Kmeans_python/fit.py (hunk 3 above)
Comment (saurav193) [reply Kmeans_python/fit.py:115]: Subjective grammatical connotation. I followed the one in sklearn kmeans source code : https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/_kmeans.py

File: Kmeans_python/fit.py (hunk 4 above)
Comment (saurav193) [reply Kmeans_python/fit.py:119]: Subjective grammatical connotation. I followed the one in sklearn kmeans source code : https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/_kmeans.py

File: Kmeans_python/fit.py (hunk 5 above)
Comment (saurav193) [reply Kmeans_python/fit.py:123]: Subjective grammatical connotation. I followed the one in sklearn kmeans source code : https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/_kmeans.py

File: Kmeans_python/fit.py (hunk 6 above)
Comment (saurav193) [reply Kmeans_python/fit.py]: I will look into this and update the code

**Review by sreejithmunthikodu — APPROVED** [review]
Looks good. The correction on inertia was much more elegantly done than suggested. Thanks, Saurav.
